source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -113,11 +102,12 @@ Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& he
 
 Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_stream) {
   if (filter_state_ == FilterState::DecodeServingFromCache) {
-    // This call was invoked by decoder_callbacks_->encodeData -- ignore it.
+    // This call was invoked during decoding by decoder_callbacks_->encodeData because a fresh
+    // cached response was found and is being added to the encoding stream -- ignore it.
     return Http::FilterDataStatus::Continue;
   }
-  if (filter_state_ == FilterState::WaitingForCacheBody) {
-    // Encoding stream stopped waiting for cached body (and trailers) to be encoded.
+  if (filter_state_ == FilterState::EncodeServingFromCache) {
```

Comment:
Why wouldn't we return Continue here? It will let they bytes flow through downstream filters and to the client, right?

Reply:
My understanding is that if we do not stop iteration, then the downstream filters will start processing the response immediately after `encodeHeaders` returns. The body may have not been fetched entirely from the cache yet. We need to fetch the complete cached body before we let the downstream filters process the response.

Actually, probably the `onBody` will be posted to the dispatcher as the last event, after all the filters finish `encodeHeaders` and `encodeData`. I think we need to stop the response encoding stream, add the cached response, then continue the encoding stream.

Reply:
I did not notice that this is `encodeData` not `encodeHeaders`, sorry.

When `encodeHeaders` returns `StopIteration`, this does not stop `encodeData` of the same filter from being called. So if the 304 response includes a body by mistake and `encodeData` is called, we need to make sure that we keep the encoding stream stopped if the cached body is not fully fetched yet.

For some reason, we cannot return `StopAllIterationAndWatermark` from `encodeHeaders`. The documentation suggests this as it states that `continueDecoding` rather than `continueDecoding/continueEncoding` must be called to continue. I also tried returning `StopAlliterationAndWatermark` and the integration test crashed.

Maybe we should actually return `StopIterationNoBuffer` here to drop any existing 304 body? I am not sure this is what it does though.

@mattklein123 can you please confirm or correct the info in this comment?

Reply:
We want downstream filters to start processing the response immediately. As soon as we have enough data to fill a packet, we want to send it on to the client, both so the client can start processing it, and so we can deallocate buffered bytes more quickly to keep our RAM usage under control.

Reply:
@toddmgreer I updated my comment, I thought the comment was about `encodeHeaders` not `encodeData`, sorry.

Reply:
I understand what you mean, but I am not sure that we can do this with `addEncodedData`. Especially that there is no way to tell it when the stream ends. What I understood from looking at the code is that we should stop iteration, add the cached data, then continue iteration.

Maybe we should use `injectDecodedDataToFilterChain` instead? But I am not exactly sure how we are supposed to use it.

Reply:
Also, `addEncodedData` does not cause `encodeData` of the same filter to be called anyway.

Reply:
It looks like to do this we will need to switch to `injectEncodedDataToFilterChain`, which is only used by the fault filter atm. Maybe postpone this to a separate PR? It wasn't introduced by this PR anyway.